### PR TITLE
Add "Build OKD-on-Fedora-CoreOS in Prow" proposal

### PR DESCRIPTION
**Pull request by LorbusChris**
OKD is the OpenShift community distribution of Kubernetes.

As part of the version 4 effort, the OKD working group has decided to target Fedora CoreOS (FCOS) as the primary base operating system for OKD control plane and worker nodes.

This document contains a proposal to add missing FCOS support to key components (`machine-config-operator` and `installer`),
create OKD artifacts in OpenShift's Prow instance from the canonical OpenShift master and release branches alongside OCP CI artifacts,
and regularily create OKD release payloads from there.

[Rendered](https://github.com/LorbusChris/enhancements/blob/okd/enhancements/okd-on-fedora-coreos.md)

### CODE REVIEW

File: enhancements/okd-on-fedora-coreos.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+A proposal to create components of an OKD variant that runs atop Fedora CoreOS and to use OpenShift's Prow instance to do so.
+
+As part of the version 4 effort, the OKD working group has agreed to target Fedora CoreOS (FCoS) as the primary base operating system for builds of version 4 of the OKD Community Distribution that is powered by the OpenShift codebase.
+
+This document proposes to incrementally add Prow CI jobs to create Fedora-based artifacts for use in OKD version 4.
+
+## Motivation
+
+Delivering continuous builds of version 4 of the OKD community variant, based on Fedora CoreOS.
```
Comment (smarterclayton) [inline enhancements/okd-on-fedora-coreos.md]: I think you can remove `community variant` and just say `OKD`

File: enhancements/okd-on-fedora-coreos.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+The broader goal is to create automated builds of all components needed for the OKD-on-FCoS effort, with a preference on running jobs in Fedora containers, and exclusively using RPMs from official Fedora repositories for Fedora CoreOS composes. Container artifacts from UBI based CI builds for OCP may be re-used, however those artifacts that come from RHEL container based builds may not be re-used to build OKD.
+
+### Non-Goals
+
+The migration to Ignition Spec v3 is not part of this proposal.
+
+## Proposal
+
+In order to create automated builds of OKD-on-FCoS, OpenShift's Prow instance shall be used to run jobs to build all required components/artifacts in Fedora containers. The definitions of these jobs will live side by side with the already existing definitions for OCP jobs in the `openshift/release` repository, on the respective operators' repos. Operators that will not go into OKD can be disregarded from this effort.
```
Comment (smarterclayton) [inline enhancements/okd-on-fedora-coreos.md]: Hrm, I don't really want to do Fedora for everything, but instead UBI (as we say below)

Comment (smarterclayton) [reply enhancements/okd-on-fedora-coreos.md]: Maybe I just misread this section

File: enhancements/okd-on-fedora-coreos.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+### Graduation Criteria
+
+Specifically, success of this proposal will be measured by the following:
+- RPMs that are dependencies of OpenShift's codebase in the base OS are built continuously in a Fedora environment: `cri-o`, `openshift-hyperkube`, `openshift-clients`, `machine-config-daemon`
+  - **Open Question**: How are builds promoted from Prow CI to FCoS release composes?
```
Comment (smarterclayton) [inline enhancements/okd-on-fedora-coreos.md]: Will those RPMs be part of FCoS?  That implies we have one FCoS release per OKD.  I would assume we would pull FCoS streams into OKD layers and simply boot from FCoS AMIs

File: enhancements/okd-on-fedora-coreos.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  - Possible alternative to Prow promotion: Automated Fedora Koji RPM builds with packit-service CI (would need implementation work)
+- Fedora CoreOS composes that contain up-to-date versions of the aforementioned dependencies are built continuously.
+- Fedora RPM based `machine-os-content` containers are built continuously
```
Comment (smarterclayton) [inline enhancements/okd-on-fedora-coreos.md]: Also, we would have a promotion gating for each OKD stream.

File: enhancements/okd-on-fedora-coreos.md (hunk 3 above)
Comment (LorbusChris) [reply enhancements/okd-on-fedora-coreos.md]: From discussions with @bgilbert and @miabbott I was under the impression we'd add those RPMs to the base OS manifest, because we need them on the bootstrap host already.

Can you clarify what you mean with OKD layers? the machine-os-content?

Comment (cgwalters) [reply enhancements/okd-on-fedora-coreos.md]: We could change the bootstrap host to pivot as well to break that dependency.

In other words, this path FCOS is pure bootimage, and we just have an "OKDFCOS" oscontainer/`machine-os-content` which contains the kubelet.

Comment (cgwalters) [reply enhancements/okd-on-fedora-coreos.md]: Having the bootstrap pivot too would be another way to fix https://github.com/openshift/installer/pull/2532

File: enhancements/okd-on-fedora-coreos.md (hunk 2 above)
Comment (LorbusChris) [reply enhancements/okd-on-fedora-coreos.md]: At least for RPM builds, I really want to avoid any potential problems caused by skew between UBI/RHEL and Fedora as buildsystem. UBI everywhere else is fine with me.

File: enhancements/okd-on-fedora-coreos.md (hunk 3 above)
Comment (LorbusChris) [reply enhancements/okd-on-fedora-coreos.md]: If we only have to produce our machine-os-content container, and not worry about the deps on the bootstrap host (i.e. in FCoS), that would be very nice indeed.

Comment (smarterclayton) [reply enhancements/okd-on-fedora-coreos.md]: It is a MASSIVE problem if we are not pivoting the bootstrap node.  That means we don't have hermetic installs based on release image, because the chosen AMI could break.

That is a serious, serious bug.

File: enhancements/okd-on-fedora-coreos.md (hunk 2 above)
Comment (LorbusChris) [reply enhancements/okd-on-fedora-coreos.md]: I clarified the passage

Comment (vrutkovs) [reply enhancements/okd-on-fedora-coreos.md]: `with a preference on running RPM build jobs in Fedora containers, and exclusively using RPMs from official Fedora repositories for Fedora CoreOS composes`

Its important to mention that binaries running on the host (hyperkube, crio, machine-config-daemon) should be built on Fedora (preferably via RPMs) to be compatible with i.e. glibc breaks

File: enhancements/okd-on-fedora-coreos.md (hunk 3 above)
Comment (cgwalters) [reply enhancements/okd-on-fedora-coreos.md]: > Is it correct that we'd then just need the MCD (as RPM) on the bootstrap host?

I think as part of this, the installer would extract the MCD binary from the release image, or alternatively run the MCD container image via podman, and do the pivot that way.  In other words, the installer would only require `podman` on the initial bootstrap host.

Comment (cgwalters) [reply enhancements/okd-on-fedora-coreos.md]: Filed https://github.com/openshift/installer/issues/2542 to track this as a dependency and ensure we have consensus from the installer team.

Comment (cgwalters) [reply enhancements/okd-on-fedora-coreos.md]: Filed https://github.com/openshift/installer/issues/2542 to track this as a dependency and ensure we have consensus from the installer team.

File: enhancements/okd-on-fedora-coreos.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+In order to create automated builds of OKD-on-FCoS, OpenShift's Prow instance shall be used to run jobs to build all required components/artifacts in Fedora containers. The definitions of these jobs will live side by side with the already existing definitions for OCP jobs in the `openshift/release` repository, or in the respective operators' repos. Operators that will not go into OKD can be disregarded from this effort.
+
+### Graduation Criteria
+
+Specifically, success of this proposal will be measured by the following:
+- RPMs that are dependencies of OpenShift's codebase in the base OS are built continuously in a Fedora environment: `cri-o`, `openshift-hyperkube`, `openshift-clients`, `machine-config-daemon`
+  - **Open Question**: How are builds promoted from Prow CI to FCoS release composes?
+  - Possible alternative to Prow promotion: Automated Fedora Koji RPM builds with packit-service CI (would need implementation work)
+- Fedora CoreOS composes that contain up-to-date versions of the aforementioned dependencies are built continuously.
+- Fedora RPM based `machine-os-content` containers are built continuously, with promotion gating for each minor version release stream.
```
Comment (smarterclayton) [inline enhancements/okd-on-fedora-coreos.md]: We need to settle where these are continuously built.  I would *prefer* that whatever RPMs are possible are built from source continuously the same way our prow CI stuff is built.  I think it's acceptable if it's in a disconnected stream that is exposed back as a pure machine-os-content if that's coming from fedora.  Alternatively, we can build the RPMs and then assemble them in prow from the latest fcos and the latest packages (which might be a short term compromise)

File: enhancements/okd-on-fedora-coreos.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+A proposal to create components of an OKD variant that runs atop Fedora CoreOS and to use OpenShift's Prow instance to do so.
+
+As part of the version 4 effort, the OKD working group has agreed to target Fedora CoreOS (FCoS) as the primary base operating system for builds of version 4 of the OKD Community Distribution that is powered by the OpenShift codebase.
+
+This document proposes to incrementally add Prow CI jobs to create Fedora-based artifacts for use in OKD version 4.
+
+## Motivation
+
+Delivering continuous builds of version 4 of OKD, based on Fedora CoreOS.
```
Comment (yuqi-zhang) [inline enhancements/okd-on-fedora-coreos.md]: Overall makes sense to me. Just curious, will these end up as separate branches of the installer/operators if customization needs to happen?

Comment (LorbusChris) [reply enhancements/okd-on-fedora-coreos.md]: This is something I think we'll decide for each component as we go forward. The general directive is to try to keep everything in one canonical master, but at least installer and mco will have their own branches in the short/mid term, i.e. until we can coordinate merging those back into master.

File: enhancements/okd-on-fedora-coreos.md (hunk 5 above)
Comment (vrutkovs) [reply enhancements/okd-on-fedora-coreos.md]: It seems there is no easy way to overlay RPMs without booting a VM. There is however a prototype which unpacks RPMs and adds binaries to the ostree, makes a new commit and updates ref.

I agree this should happen on Prow (so that Fedora would not have to maintain additional stream of FCOS), use crio RPMs from fedora upstream and overlay binaries from MCD / Origin images, skipping RPMs phase if necessary

Comment (LorbusChris) [reply enhancements/okd-on-fedora-coreos.md]: +1 to Prow.
I can also live with overlaying just the binaries from the Prow-built rpms. Should we use your `machine-os-content-builder` for that? I like the approach.

Comment (vrutkovs) [reply enhancements/okd-on-fedora-coreos.md]: That's just a prototype, its not really compatible with Prow just yet. Clayton has a better working solution

**Comment by LorbusChris** [comment]
I've updated the proposal to reflect the current status and the work left to get us to OKD 4 GA.

/cc @smarterclayton @miabbott @ashcrow @runcom @darkmuggle @vrutkovs @sdodson @crawford @abhinavdahiya @cgwalters

File: enhancements/okd-on-fedora-coreos.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+### 4: Installer Changes
+
+These changes are optional for the immediate goal of releasing OKD 4 GA. They should however be strongly considered in order to facilitate dropping the OKD fork of installer rather sooner than later.
+
+#### Add fcos.json
+
+Support for an `fcos.json` to exist alongside `rhcos.json` for OS image references.
+
+This could be enabled either via a build-time flag to produce seperate OKD installer artifacts, or an `fcos` or `okd` run-time flag to pass to the installer for it to provision Fedora CoreOS instead of RHEL CoreOS hosts (in that case it would default to contents of `rhcos.json`).
```
Comment (vrutkovs) [inline enhancements/okd-on-fedora-coreos.md:129]: I'd prefer the build-time flag, similar to what libvirt approach:

* [ ] Rework existing installer to have all ignition operators in a single file (https://github.com/openshift/installer/pull/3248)
* [ ] Create Ignition 3 implementation of the same functions, hidden under `okd` build flag
* [ ] Build `installer:okd` image with this flag enabled
* [ ] Add `isOKD` internal setting and conditionally enable okd-specific features:
  * use `fcos.json` instead of `rhcos.json`
  * allow empty pull secret
  * restrict to community operators
* [ ] Enable libvirt provider in OKD installer

Comment (LorbusChris) [reply enhancements/okd-on-fedora-coreos.md:129]: +1 from me. @abhinavdahiya, WDYT?

Comment (smarterclayton) [reply enhancements/okd-on-fedora-coreos.md:129]: Why not rename rhcos.json to machine-os.json and make that the build time replacement instead?

Branches on “is okd” raise red flags for me - the imply OKD isn’t the same source.

The restriction to community operators can be handled dynamically in many cases (does the payload include the operator).

The pull secret is going to be relevant for some deployments no matter what - perhaps instead of making the check conditional we instead check whether the pull of the release image requires auth as a gate during pre validation (which has the side benefit of verifying the credentials)

Comment (cgwalters) [reply enhancements/okd-on-fedora-coreos.md:129]: Bikeshed: `bootimage.json`

Comment (vrutkovs) [reply enhancements/okd-on-fedora-coreos.md:129]: > Why not rename rhcos.json to machine-os.json and make that the build time replacement instead?

Sounds good to me

> Branches on “is okd” raise red flags for me - the imply OKD isn’t the same source.

We still need to branch on few options here - i.e. default SDN, Ignition v3 etc. `IsOKD` is not exposed externally, cannot be changed via config and being set on build time

> The restriction to community operators can be handled dynamically in many cases (does the payload include the operator).

So instead of installer injecting operators we'd have to have a setting other operators would read? 

> we instead check whether the pull of the release image requires auth as a gate during pre validation

This check would happen in the installer, right?

File: enhancements/okd-on-fedora-coreos.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+With the MCO's ability to process both spec v2 and v3 configs, switching the installer over to spec v3 is technically possible.
+
+It is imperative that the timing of this change is aligned with RHEL CoreOS images switching to an Ignition v2.x binary that supports spec v3.
+
+Once this migration is complete, it will be possible for OKD to use an installer built off of the same sources.
+
+### Graduation Criteria
+
+Specifically, success of this proposal will be measured by the following:
+- Fedora RPM based `machine-os-content` containers are built continuously, with promotion gating for each minor version release stream.
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply enhancements/okd-on-fedora-coreos.md:152]: `stable` stream of Fedora CoreOS (currently we're pivoting to `testing-devel`)

**Comment by LorbusChris** [comment]
Are there any blockers to this?

**Comment by ashcrow** [comment]
Today in a call it was noted there are a few updates to this enhancement which should show up soon.

**Comment by jomeier** [comment]
Hi. OKD 4 is GA. Does this open PR mean that the team has not agreed to this Proposal?

**Comment by LorbusChris** [comment]
No. I'll be updating/adding a few details on this soon.

**Comment by russellb** [comment]
> No. I'll be updating/adding a few details on this soon.

still planning to update this one?

**Comment by LorbusChris** [comment]
@russelb the path to merge installer master and fcos branches is not quite clear, yet, and probably won't happen for another release or two. I'm banking on #201 to make that easier.

For this proposal here, maybe it makes sense to split out the installer part and merge the rest that's already been implemented?

**Comment by vrutkovs** [comment]
Some more details we need to cover here:
* OKD requires custom machine-os-content based on FCOS, hypercube, mcd and crio.
  Currently it gets [built by a massive periodic job](https://github.com/openshift/release/blob/b79f5011557cfd14532beaa841c6e7076de5cc25/ci-operator/jobs/openshift/release/openshift-release-release-4.6-periodics.yaml#L653-L983). Instead we should create a new repository for this (example - https://github.com/vrutkovs/machine-os-content-builder) with OWNERS, approvals and rehearsals. The repo would also contain custom OKD configuration - machine-configs which enable OVS OS extensions, custom OperatorHub configuration etc.
This repo would run periodic builds and promotions.
* Installer would produce okd-specific image (OVN enabled by default, FCOS artifacts are used instead of RHCOS etc.) Current OKD fork implements this via build tag ([here](https://github.com/openshift/installer/commit/aea0fa3edf0bbe4a5fbd6f174f7d2e9648604e82) and [here](https://github.com/openshift/installer/commit/69dfa57ba2a7571523ed44814b7e6a97607c9a17), [OVN in OKD by default](https://github.com/openshift/installer/commit/30106cd70d2d39b416acc05e5918194e05789dfb)), however a different method may be used.
* https://github.com/openshift/okd used to triage and track OKD-specific bugs

File: enhancements/okd-on-fedora-coreos.md [hunk 9]
```diff
@@ -32,128 +32,58 @@ superseded-by:
...
+As part of the version 4 effort, the OKD working group has decided to target Fedora CoreOS (FCOS) as the primary base operating system for OKD control plane and worker nodes.
 
-This document contains a proposal to create all OKD artifacts in OpenShift's Prow instance from the canonical OpenShift master and release branches alongside OCP artifacts.
-
-Eventually, container images for all components will be shared between OKD and OCP, with the exception of the machine-os-content container image which will be based on Fedora CoreOS content, and any container images that are based on RHEL images in OCP. These will be rebuilt on either a UBI, Fedora or CentOS base for OKD.
-
-A lot of this is already done. What's left is the un-forking of Machine Config Operator and (optionally) Installer repository and building variant-agnostic artifacts of them, as well as changing the Samples Operator to not go degraded on OKD clusters.
+This document contains a proposal to add missing FCOS support to key components (`machine-config-operator` and `installer`),
```
Comment (vrutkovs) [inline enhancements/okd-on-fedora-coreos.md:41]: We've also created OKD variants for `samples` (contains CentOS and UBI) and `marketplace` (community operators index can be pulled without a special pull secret). These are not critical and not sure if its worth mentioning those

File: enhancements/okd-on-fedora-coreos.md [hunk 10]
```diff
@@ -32,128 +32,58 @@ superseded-by:
...
-
-Only a limited subset of the Ignition config fields is supported by the MCS (i.e. config that describes the on-disk state like files and systemd units).
-
-In OCP this on-disk state representation used within the Installer and the MCS is currently encoded in Ignition spec v2.2 config types, while v3.0 config types are used in OKD. To account for the different spec types, OKD's Installer binary as well as OKD's Machine Config Operator (MCO) container image are currently built from forked codebases.
-
-Long-term maintainance of those forks is not wanted. The MCO fork represents an OKD 4 GA release blocker. The installer fork, while not blocking GA, should be dissolved after OCP migrates to Ignition config spec v3.
+The OKD Working Group wants to deliver a community variant of OpenShift on top of Fedora CoreOS\
```
Comment (vrutkovs) [inline enhancements/okd-on-fedora-coreos.md:47]:
```suggestion
The OKD Working Group wants to deliver a community variant of OpenShift on top of Fedora CoreOS
```

File: enhancements/okd-on-fedora-coreos.md [hunk 11]
```diff
@@ -32,128 +32,58 @@ superseded-by:
...
+that is released off of the latest OpenShift stable release branch in a rolling manner,
+and provids the ability to update. An update release payload for OKD encapsulates a new release of the underlying operating system
```
Comment (vrutkovs) [inline enhancements/okd-on-fedora-coreos.md:49]:
```suggestion
and provides the ability to update. An update release payload for OKD encapsulates a new release of the underlying operating system
```

File: enhancements/okd-on-fedora-coreos.md [hunk 12]
```diff
@@ -32,128 +32,58 @@ superseded-by:
...
-To facilitate the migration, all functions requiring Ignition types will be abstracted into one package. This is implemented by this PR: https://github.com/openshift/installer/pull/3248
-
-With the MCO's ability to process both spec v2 and v3 configs, switching the installer over to spec v3 is technically possible.
-
-It is imperative that the timing of this change is aligned with RHEL CoreOS images switching to an Ignition v2.x binary that supports spec v3.
-
-Once this migration is complete, it will be possible for OKD to use an installer built off of the same sources.
+This is a proposal to:
+- Create all OKD artifacts in OpenShift's Prow instance from the canonical OpenShift master branch and the latest release branch for testing,
+and release OKD payloads built in Prow from the latest release branch regularly, e.g. every 3 weeks.
```
Comment (vrutkovs) [inline enhancements/okd-on-fedora-coreos.md:65]:
```suggestion
and release OKD payloads built in Prow from the latest release branch regularly, e.g. every 2 weeks.
```
we're at two week schedule now, lets put that

File: enhancements/okd-on-fedora-coreos.md [hunk 13]
```diff
@@ -32,128 +32,58 @@ superseded-by:
...
-
-To facilitate the migration, all functions requiring Ignition types will be abstracted into one package. This is implemented by this PR: https://github.com/openshift/installer/pull/3248
-
-With the MCO's ability to process both spec v2 and v3 configs, switching the installer over to spec v3 is technically possible.
-
-It is imperative that the timing of this change is aligned with RHEL CoreOS images switching to an Ignition v2.x binary that supports spec v3.
-
-Once this migration is complete, it will be possible for OKD to use an installer built off of the same sources.
+This is a proposal to:
+- Create all OKD artifacts in OpenShift's Prow instance from the canonical OpenShift master branch and the latest release branch for testing,
```
Comment (vrutkovs) [inline enhancements/okd-on-fedora-coreos.md:64]: We create OKD from all active branching (even 4.3 nightlies are being produced). However we'll promote to stable only the latest supported branch (`release-4.6` at the time). Not sure if that needs to be clarified

File: enhancements/okd-on-fedora-coreos.md [hunk 14]
```diff
@@ -32,128 +32,58 @@ superseded-by:
...
+- [x] OKD build and release jobs are added to `openshift/release`
+- [x] Fedora RPM- and FCOS ostree-based `machine-os-content` container images are built continuously from the [`openshift/okd-machine-os`](https://github.com/openshift/okd-machine-os/) repository, with promotion gating for each minor version release stream.
+- [x] Mirror OKD release payloads from the internal registry to quay.io  
+- [x] Support and end-to-end CI testing is added for running `machine-config-operator`, and the `machine-config-daemon` on FCOS-based OKD clusters.
+- [x] `machine-config-operator` sources do not require a fork to be built for OKD
+- [x] Support and end-to-end CI testing is added for building OKD installer binaries
+- [ ] `openshift-install` sources do not require a fork to be built for OKD: ([Pull Request](https://github.com/openshift/installer/pull/4453))
+- [x] All container images that are part of operators shipped by `cluster-samples-operator` are continuously available as either Fedora, CentOS or UBI-based container images.
+- [ ] All container images that are part of `baremetal-operator` are continuously available as either Fedora, CentOS or UBI-based container images.
```
Comment (vrutkovs) [inline enhancements/okd-on-fedora-coreos.md:84]: Tracking bug - https://github.com/openshift/okd/issues/197

**Review by vrutkovs — APPROVED** [review]
Looks good, found few typos

File: enhancements/okd-on-fedora-coreos.md (hunk 13 above)
Comment (LorbusChris) [reply enhancements/okd-on-fedora-coreos.md:64]: May be it should be `master branch and at least the latest release branch`.

In the future, we could possible remove the building of versions before that to save CI costs - there's nobody consuming these artifacts after all.

Comment (vrutkovs) [reply enhancements/okd-on-fedora-coreos.md:64]: >master branch and at least the latest release branch

This sounds good.

> In the future, we could possible remove the building of versions before that to save CI costs - there's nobody consuming these artifacts after all.

IIUC we're keeping last 5 artifacts, regardless of the age, so it shouldn't use too much resources. Promotion jobs are still running once a day for "old stable" streams, so it shouldn't affect CI too much

**Comment by LorbusChris** [comment]
@staebler I've added you to the reviewers from the Installer side.

Tl;dr the largest remaining piece of this is https://github.com/openshift/installer/pull/4453.

**Comment by mjturek** [comment]
I'm interested in building for other architectures. I'm tasked with investigating OKD support for ppc64le.

After a conversation @vrutkovs and @LorbusChris, it seems the goal would be to do nightly builds of OKD on ppc64le in cadence with x86_64. This means we would need infrastructure where we could build these containers, and a release controller.

Can others weigh in on where we may be able to do this? Any input is welcome and appreciated!

**Comment by cgwalters** [comment]
I think we'd want to ask either Fedora infra or CentOS infra about capacity for this - that's where the community ppc64le/s390x hardware is as far as I know.

**Comment by derekwaynecarr** [comment]
the enhancement should be a statement of intent or direction, and so holding this out while the installer change is reviewed seems unnecessary.  are there any further updates required, or we good to merge and iterate?

**Comment by LorbusChris** [comment]
@derekwaynecarr I think this is good to merge.

Setting up OKD builds on alt arches is not part of this proposal, and can be done in a separate one if needed.

**Comment by russellb** [comment]
Merging based on the last comments from @derekwaynecarr and @LorbusChris 

/approve
/lgtm

**Comment by LorbusChris** [comment]
I've fixed the linter issues.

/hold cancel

**Comment by darkmuggle** [comment]
This enhancement was not approved and it's not implementable. Stakeholders (Installer, MCO, CoreOS, among others) have not agreed, nor have their reviews been obtained.

**Comment by russellb** [comment]
> This enhancement was not approved and it's not implementable. Stakeholders (Installer, MCO, CoreOS, among others) have not agreed, nor have their reviews been obtained.

This was merged primarily as a starting point to document the current reality (OKD does already exist ...) and make it easier to iterate on the doc.  You should not interpret this as any big new decision.  I've spoken with @vrutkovs about some updates needed here to better capture status and open questions and have asked him to tag you on the update.
